manager/state/raft/storage/nacl.go

```diff
@@ -0,0 +1,74 @@
+package storage
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/nacl/secretbox"
```

Comment:
AES-GCM has a number of issues that are relevant, depending on context. Some of them are irrelevant (e.g. security of truncated tags), some of them are always relevant (e.g. limitations on total amount encrypted, short nonces implying implicit-to-the-protocol nonces). The nonce issue is the biggest one for above-Raft encryption: you need to agree on who gets to use which nonce before you can produce a ciphertext to propose for entering in the log. You can't use (term, index), because you need the nonce before you can produce the ciphertext, so you are necessarily producing duplicate-nonce ciphertexts. You can't pick random nonces, because GCM nonces are short. So, you need to run the nonce through the Raft log, or use a stronger primitive that does not have those limitations.

This is less of a problem for on-disk Raft log encryption, because you have an obvious nonce there: you _can_ use the (term, index) since there's only one encryption per (plaintext, key) pair, iff the key is per-node.

(GCM in the context of TLS doesn't have this protocol, because TLS is already a record protocol, so you have a natural, implicit nonce. [That doesn't stop people from messing it up](https://eprint.iacr.org/2016/475.pdf), but that's neither here nor there :))